src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,1840 @@
+/** -*- C++ -*- *****************************************************\
+
+Module: Constraint generation from string function calls
+        for the PASS algorithm (see the PASS paper at HVC'13)
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#include <solvers/refinement/string_constraint_generator.h>
+#include <util/unicode.h>
+#include <ansi-c/string_constant.h>
+#include <solvers/floatbv/float_bv.h>
+#include <java_bytecode/java_types.h>
+#include <util/pointer_predicates.h>
+
+constant_exprt string_constraint_generatort::constant_char(int i)
+{
+  switch(language)
+    {
+    case C :
+      return constant_exprt(integer2binary(i,STRING_SOLVER_CHAR_WIDTH),refined_string_typet::char_type());
+      break;
+    case JAVA : return constant_exprt(integer2binary(i,JAVA_STRING_SOLVER_CHAR_WIDTH),refined_string_typet::java_char_type());
+      break;
+    default: assert(false);
+    }
+}
+
+
+constant_exprt string_constraint_generatort::constant_unsigned(int i, size_t width)
+{ return constant_exprt(integer2binary(i,width),unsignedbv_typet(width)); }
+
+constant_exprt string_constraint_generatort::constant_signed(int i, size_t width)
+{ return constant_exprt(integer2binary(i,width),signedbv_typet(width)); }
+
+void string_constraint_generatort::check_char_type(const exprt & str)
+{
+  if(language == C)
+    assert(refined_string_typet::is_c_string_type(str.type()));
+  else
+    if(language == UNKNOWN)
+      {
+	if(refined_string_typet::is_c_string_type(str.type()))
+	  language = C;
+	else
+	  language = JAVA;
+      }
+
+}
```

Comment:
I did not find a very good way to do that. 
My attempt at making it cleaner is here https://github.com/diffblue/cbmc/pull/341/commits/3c1daf04eed37dffe7e3f062c5bdcd7b900ba443